File: interface/src/main/java/eu/nyerel/panda/monitoringresult/calltree/CallTreeNodeDuration.java
package eu.nyerel.panda.monitoringresult.calltree;

import java.io.Serializable;

/**
 * @author Rastislav Papp (rastislav.papp@gmail.com)
 */
public class CallTreeNodeDuration implements Serializable {

    private static final long serialVersionUID = 1L;

    private long total;
    private long self;
    private long children;
    private double percentage;

    public CallTreeNodeDuration add(CallTreeNodeDuration duration) {
        CallTreeNodeDuration sum = new CallTreeNodeDuration();
        sum.setTotal(total + duration.total);
        sum.setSelf(self + duration.self);
        sum.setChildren(children + duration.children);
        sum.setPercentage(percentage + duration.percentage);
        return sum;
    }

    public CallTreeNodeDuration divide(int number) {
        CallTreeNodeDuration divided = new CallTreeNodeDuration();
        if (number > 0) {
            divided.setTotal(total / number);
            divided.setSelf(self / number);
            divided.setChildren(children / number);
            divided.setPercentage(percentage / number);
        }
        return divided;
    }

    public long getTotal() {
        return total;
    }

    public void setTotal(long total) {
        this.total = total;
    }

    public long getSelf() {
        return self;
    }

    public void setSelf(long self) {
        this.self = self;
    }

    public long getChildren() {
        return children;
    }

    public void setChildren(long children) {
        this.children = children;
    }

    public double getPercentage() {
        return percentage;
    }

    public void setPercentage(double percentage) {
        this.percentage = percentage;
    }

    @Override
    public String toString() {
        return "total: " + formatTotal() + ", self: " + formatSelf();
    }

    public String formatTotal() {
        return total + " ms (" + formatPercentage(percentage) + ")";
    }

    private String formatPercentage(double percentage) {
        return Math.round(Math.round((100 * percentage))) + "%";
    }

    public String formatSelf() {
        long percentage = total == 0 ? 0 : self / total;
        return self + " ms (" + formatPercentage(percentage) + ")";
    }

}

File: interface/src/main/kotlin/eu/nyerel/panda/monitoringresult/calltree/CallTreeNodeDuration.kt
package eu.nyerel.panda.monitoringresult.calltree

import java.io.Serializable

/**
 * @author Rastislav Papp (rastislav.papp@gmail.com)
 */
class CallTreeNodeDuration : Serializable {

    var total: Long = 0
    var self: Long = 0
    var children: Long = 0
    var percentage: Double = 0.toDouble()

    fun add(duration: CallTreeNodeDuration): CallTreeNodeDuration {
        val sum = CallTreeNodeDuration()
        sum.total = total + duration.total
        sum.self = self + duration.self
        sum.children = children + duration.children
        sum.percentage = percentage + duration.percentage
        return sum
    }

    fun divide(number: Int): CallTreeNodeDuration {
        val divided = CallTreeNodeDuration()
        if (number > 0) {
            divided.total = total / number
            divided.self = self / number
            divided.children = children / number
            divided.percentage = percentage / number
        }
        return divided
    }

    override fun toString(): String {
        return "total: " + formatTotal() + ", self: " + formatSelf()
    }

    fun formatTotal(): String {
        return total.toString() + " ms (" + formatPercentage(percentage) + ")"
    }

    private fun formatPercentage(percentage: Double): String {
        return Math.round(Math.round(100 * percentage).toFloat()).toString() + "%"
    }

    fun formatSelf(): String {
        val percentage = if (total == 0L) 0 else self / total
        return self.toString() + " ms (" + formatPercentage(percentage.toDouble()) + ")"
    }

    companion object {

        private const val serialVersionUID = 1L
    }

}
